core/src/main/java/com/sequenceiq/cloudbreak/reactor/CollectDownscaleCandidatesHandler.java

```diff
@@ -40,17 +50,35 @@ public void accept(Event<CollectDownscaleCandidatesRequest> event) {
         CollectDownscaleCandidatesResult result;
         try {
             Stack stack = stackService.getByIdWithLists(request.getStackId());
-            Set<String> hostNames;
-            if (request.getHostNames() == null || request.getHostNames().isEmpty()) {
-                hostNames = ambariDecommissioner.collectDownscaleCandidates(stack, request.getHostGroupName(), request.getScalingAdjustment());
-            } else {
-                hostNames = request.getHostNames();
-                ambariDecommissioner.verifyNodeCount(stack, stack.getCluster(), hostNames.stream().findFirst().get());
-            }
-            result = new CollectDownscaleCandidatesResult(request, hostNames);
+            result = verifyRequestAndCreateResult(request, stack);
         } catch (Exception e) {
             result = new CollectDownscaleCandidatesResult(e.getMessage(), e, request);
         }
         eventBus.notify(result.selector(), new Event<>(event.getHeaders(), result));
     }
+
+    private CollectDownscaleCandidatesResult verifyRequestAndCreateResult(CollectDownscaleCandidatesRequest request, Stack stack) throws CloudbreakException {
+        CollectDownscaleCandidatesResult result;
+        if (isFqdnMissing(request) && Objects.isNull(request.getScalingAdjustment())) {
+            LOGGER.error("No FQDN and scaling adjustment was found in the CollectDownscaleCandidatesRequest.");
+            result = new CollectDownscaleCandidatesResult(ERROR_STATUS_REASON, new CloudbreakException(ERROR_STATUS_REASON), request);
+        } else if (isFqdnMissing(request)) {
+            LOGGER.warn("No FQDN was found in the CollectDownscaleCandidatesRequest.");
+            Set<String> hostNames = ambariDecommissioner.collectDownscaleCandidates(stack, request.getHostGroupName(), request.getScalingAdjustment());
+            result = new CollectDownscaleCandidatesResult(request, hostNames);
+        } else {
+            Set<String> hostNames = request.getHostNames();
+            ambariDecommissioner.verifyNodeCount(stack, stack.getCluster(), hostNames.stream().findFirst().get());
+            result = new CollectDownscaleCandidatesResult(request, hostNames);
+        }
+        return result;
+    }
+
+    private boolean isFqdnMissing(CollectDownscaleCandidatesRequest request) {
+        try {
+            return CollectionUtils.isEmpty(request.getHostNames()) || StringUtils.isEmpty(request.getHostNames().stream().findFirst().get());
+        } catch (NullPointerException e) {
```

Comment:
Control by exception is a bad practice because it is expensive and hard to maintenance